From 02c752c017d213a734c1a9e6590d7b500b7733ea Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Fri, 29 Aug 2014 21:18:44 -0400
Subject: [net] sunrpc: fix potential race between setting use_gss_proxy and the upcall rpc_clnt

Message-id: <1409347125-11495-6-git-send-email-bfields@redhat.com>
Patchwork-id: 89917
O-Subject: [Patch RHEL7 5/6] sunrpc: fix potential race between setting use_gss_proxy and the upcall rpc_clnt
Bugzilla: 1117914
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Jeff Layton <jlayton@redhat.com>

An nfsd thread can call use_gss_proxy and find it set to '1' but find
gssp_clnt still NULL, so that when it attempts the upcall the result
will be an unnecessary -EIO.

So, ensure that gssp_clnt is created first, and set the use_gss_proxy
variable only if that succeeds.

Signed-off-by: Jeff Layton <jlayton@redhat.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
(cherry picked from commit a92e5eb1103341e985a575e48e26f87fbb9b1679)
Bugzilla 1117914

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/sunrpc/auth_gss/svcauth_gss.c b/net/sunrpc/auth_gss/svcauth_gss.c
index 8c05037..3489cd4 100644
--- a/net/sunrpc/auth_gss/svcauth_gss.c
+++ b/net/sunrpc/auth_gss/svcauth_gss.c
@@ -1317,10 +1317,10 @@ static ssize_t write_gssp(struct file *file, const char __user *buf,
   return res;
  if (i != 1)
   return -EINVAL;
- res = set_gss_proxy(net, 1);
+ res = set_gssp_clnt(net);
  if (res)
   return res;
- res = set_gssp_clnt(net);
+ res = set_gss_proxy(net, 1);
  if (res)
   return res;
  return count;
-- 
1.7.1